tracer/src/Datadog.Trace/Debugger/Symbols/SymbolBatchUploadApi.cs

```diff
@@ -0,0 +1,133 @@
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Failed to initialize event {Event}", @event);
+                return default;
+            }
+        }
+
+        public static SymbolBatchUploadApi Create(IApiRequestFactory apiRequestFactory, IDiscoveryService discoveryService)
+        {
+            return new SymbolBatchUploadApi(apiRequestFactory, discoveryService);
+        }
+
+        public async Task<bool> SendBatchAsync(ArraySegment<byte> symbols)
+        {
+            if (Volatile.Read(ref _endpoint) is not { } endpoint)
+            {
+                Log.Warning("Failed to upload symbol: symbol endpoint not yet retrieved from discovery service");
+                return false;
+            }
+
+            if (_event == default || _event.Count == 0)
+            {
+                Log.Error("Failed to initialize event");
+                return default;
+            }
+
+            var uri = _apiRequestFactory.GetEndpoint(endpoint);
+            var request = _apiRequestFactory.Create(uri);
+            var multipartRequest = (IMultipartApiRequest)request;
+
+            const int maxRetries = 3;
+            int retries = 0;
+
+            while (retries < maxRetries)
+            {
+                using var response = await multipartRequest.PostAsync(
+                                         new MultipartFormItem("symbols", MimeTypes.Json, null, symbols),
+                                         new MultipartFormItem("event", MimeTypes.Json, "event", _event)).
+                                                            ConfigureAwait(false);
+
+                if (response.StatusCode is >= 200 and <= 299)
+                {
+                    return true;
+                }
+
+                retries++;
+                var content = await response.ReadAsStringAsync().ConfigureAwait(false);
+
+                if (ShouldRetry(response.StatusCode))
+                {
+                    await Task.Delay(GetDelayTime(retries)).ConfigureAwait(false);
+                }
+                else
+                {
+                    Log.Error<int, string>("Failed to upload symbol with status code {StatusCode} and message: {ResponseContent}", response.StatusCode, content);
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool ShouldRetry(int statusCode)
+        {
+            int[] statusCodesToRetry = { 408, 425, 429, 503, 504 };
+            return statusCodesToRetry.Any(code => code == statusCode);
+        }
+
+        private static TimeSpan GetDelayTime(int retryAttempt)
+        {
+            const int baseDelayMs = 250;
+            int maxDelayMs = baseDelayMs * (int)Math.Pow(2, retryAttempt);
+
+            Random random = new Random();
+            int delayMs = random.Next(baseDelayMs, maxDelayMs);
+
+            return TimeSpan.FromMilliseconds(delayMs);
```

Comment:
I was wondering if we really need to allocate a 'Random' here each time, and if we need randomness at all? But as a more general comment, should we try to just reuse the same code for [retry logic that appear in other places in the code base](https://github.com/search?q=retry+repo%3ADatadog%2Fdd-trace-dotnet+language%3AC%23&type=code&expanded_query=retry+repo%3ADatadog%2Fdd-trace-dotnet&l=C%23)?